plutus-tx-plugin/src/PlutusTx/Compiler/Expr.hs

```diff
@@ -422,6 +431,43 @@ hoistExpr var t =
                     (Set.map LexName deps)
                 pure $ PIR.mkVar () var'
 
+mkTrace
+    :: (PLC.Contains uni T.Text)
+    => PLC.Type PLC.TyName uni ()
+    -> T.Text
+    -> PIRTerm uni PLC.DefaultFun
+    -> PIRTerm uni PLC.DefaultFun
+mkTrace ty str v =
+    PLC.mkIterApp
+        ()
+        (PIR.TyInst () (PIR.Builtin () PLC.Trace) ty)
+        [PLC.mkConstant () str, v]
+
+traceInside ::
+    T.Text
+    -> PIR.Name
+    -> PIRTerm PLC.DefaultUni PLC.DefaultFun
+    -> PLC.Type PIR.TyName PLC.DefaultUni ()
+    -> PIRTerm PLC.DefaultUni PLC.DefaultFun
+traceInside varName lamName (LamAbs () n t body) (PLC.TyFun () _dom cod) =
+    -- when t = \x -> body, => \x -> traceInside body
+    LamAbs () n t (traceInside varName lamName body cod)
+traceInside _ _ LamAbs{} _ =
+    error "traceInside: type mismatched. It should be a function type."
+traceInside varName lamName e ty =
+    let defaultUnitTy = PLC.TyBuiltin () (PLC.SomeTypeIn PLC.DefaultUniUnit)
+        defaultUnit = PIR.Constant () (PLC.someValueOf PLC.DefaultUniUnit ())
+    in
+    --(trace @(() -> c) "entering f" (\() -> trace @c "exiting f" body) ())
+        PIR.Apply
+            ()
+            (mkTrace
+                (PLC.TyFun () defaultUnitTy ty) -- ()-> ty
+                ("entering " <> varName)
+                -- \() -> trace @c "exiting f" e
+                (LamAbs () lamName defaultUnitTy (mkTrace ty ("exiting "<>varName) e)))
```

Comment:
I'd probably just make this function monadic so you can generate the name here, but it's also fine to pass it in.

Reply:
As a FPer, I tend to avoid monads when I can. ;)